factcast-snapshotcache-jdbc/src/main/java/org/factcast/core/snap/jdbc/JdbcSnapshotCache.java

```diff
@@ -0,0 +1,173 @@
+            + " INSERT VALUES (_key, _uuid, _last_fact_id, _bytes, _compressed, _last_accessed)";
+
+    deleteStatement = "DELETE FROM " + tableName + " WHERE key = ? AND uuid = ?";
+
+    boolean snapTableExists = doesTableExist(tableName);
+
+    if (!snapTableExists) {
+      throw new IllegalStateException("Snapshots table does not exist: " + tableName);
+    } else {
+      validateColumns(tableName);
+    }
+
+    if (properties.getDeleteSnapshotStaleForDays() > 0) {
+      createTimer()
+          .scheduleAtFixedRate(
+              new StaleSnapshotsTimerTask(
+                  dataSource, tableName, properties.getDeleteSnapshotStaleForDays()),
+              0,
+              TimeUnit.DAYS.toMillis(1));
+    } else {
+      log.info("Scheduled Snapshot cleanup is disabled");
+    }
+  }
+
+  // exists in order to be able to hook into during unit tests
+  protected Timer createTimer() {
+    // we want to be able to identify it and do not want a dangling service on shutdown
+    return new Timer("JdbcSnapshotCache", true);
+  }
+
+  @SneakyThrows
+  public boolean doesTableExist(String tableName) {
+    try (Connection connection = dataSource.getConnection();
+        ResultSet rs =
+            connection.getMetaData().getTables(null, null, tableName, new String[] {"TABLE"})) {
+
+      return Boolean.TRUE.equals(rs.next());
+    }
+  }
+
+  @SneakyThrows
+  public void validateColumns(String tableName) {
+    try (Connection connection = dataSource.getConnection();
+        ResultSet columns = connection.getMetaData().getColumns(null, null, tableName, null)) {
+
+      Set<String> columnsSet =
+          Sets.newHashSet("key", "uuid", "last_fact_id", "bytes", "compressed", "last_accessed");
+      while (columns.next()) {
+        String columnName = columns.getString("COLUMN_NAME");
+
+        columnsSet.remove(columnName);
+      }
+      if (!columnsSet.isEmpty()) {
+        throw new IllegalStateException(
+            "Snapshot table schema is not compatible with Factus. Missing columns: " + columnsSet);
+      }
+    }
+  }
+
+  @Override
+  @SneakyThrows
+  public @NonNull Optional<Snapshot> getSnapshot(@NonNull SnapshotId id) {
+    try (Connection connection = dataSource.getConnection();
+        PreparedStatement statement = connection.prepareStatement(queryStatement)) {
+      statement.setString(1, id.key());
+      statement.setString(2, id.uuid().toString());
+      try (ResultSet resultSet = statement.executeQuery()) {
+        if (resultSet.next()) {
+          SnapshotId snapshotId =
+              SnapshotId.of(resultSet.getString(1), UUID.fromString(resultSet.getString(2)));
+          Snapshot snapshot =
+              new Snapshot(
+                  snapshotId,
+                  UUID.fromString(resultSet.getString(3)),
+                  resultSet.getBytes(4),
+                  resultSet.getBoolean(5));
+
+          // update last accessed
+          setSnapshot(snapshot);
```

Comment:
is it really a good idea to send whole bytearray back to the database, just to get the timestamp updated?!